### DISCUSSION TITLE
How to  create content items from a bulk csv file import

### DISCUSSION THREAD

**Discussion by ricocsharp** [OrchardCMS/OrchardCore#12203] [General]
Hi everyone,  there is a way to make a bulk importation of  data from a csv file and create contentitems with it?

**Comment by Skrypt** [comment]
It is possible but with custom code and by using the ContentManager service.

**Comment by lampersky** [comment]
@ricocsharp also you could use **Workflows Module** if this csv can be fetched using **HttpRequestTask**, or another idea is to use **Deployment module** and transform this csv into [recipe file](https://docs.orchardcore.net/en/dev/docs/reference/modules/Recipes/#content-step). But, as @Skrypt said, custom code is the easiest one.

**Reply by lampersky** [reply]
@ricocsharp here it is an example, how you can create recipe manually, just add reading CSV file and create all necessary things like ContentType, ContentPart and you can import data this way:

```
        public class MyPart : ContentPart
        {
            public MyPart() { }
            public TextField Name { get; set; }
            public MediaField Image { get; set; }
        }
```

controller method:

```
        [HttpGet("example-recipe")]
        [DeleteFileResultFilter]
        public async Task<IActionResult> Example()
        {
            using var fileBuilder = new TemporaryFileBuilder();

            var archiveFileName = fileBuilder.Folder + ".zip";
            var recipeDescriptor = new RecipeDescriptor();
            var deploymentPlanResult = new DeploymentPlanResult(fileBuilder, recipeDescriptor);

            var data = new JArray();

            for (var i = 0; i < 100; i++)
            {
                var contentItem = new ContentItem();
                contentItem.ContentItemId = IdGenerator.GenerateId();
                contentItem.ContentItemVersionId = IdGenerator.GenerateId();
                contentItem.DisplayText = $"item {i}";
                contentItem.Author = "lampersky";
                contentItem.ContentType = "MyContentType";
                contentItem.Latest = true;
                contentItem.Published = true;
                contentItem.CreatedUtc = DateTime.UtcNow;
                contentItem.Alter<MyPart>(a =>
                {
                    a.Image = new MediaField { Paths = new string[] { $"test/image_{i}.png" } };
                    a.Name = new TextField { Text = $"example name {i}" };
                });

                var objectData = JObject.FromObject(contentItem);
                objectData.Remove(nameof(ContentItem.Id));
                data.Add(objectData);
            }

            deploymentPlanResult.Steps.Add(new JObject(
                new JProperty("name", "Content"),
                new JProperty("data", data)
            ));

            await deploymentPlanResult.FinalizeAsync();

            ZipFile.CreateFromDirectory(fileBuilder.Folder, archiveFileName);

            return new PhysicalFileResult(archiveFileName, "application/zip") { FileDownloadName = "example.zip" };
        }
```

this is my generated recipe:

![image](https://user-images.githubusercontent.com/6403130/185407333-b9b4757f-cd19-4e7f-a6f9-7eb689dbc760.png)

**Reply by ricocsharp** [reply]
@lampersky thanks a lot I all your ideas helped me a lot I will share the way I'm using to acomplish this in a short while

**Comment by ricocsharp** [comment]
well thanks @lampersky  and @Skrypt  a lot for your ideas in fact both will be very usefull in different moments of my project do U guys have any recomendation of examples or videos on both opproaches?

**Reply by lampersky** [reply]
when it comes to create ContentItems via Workflows Module you can have a look at my WordPress importer example project
 https://github.com/lampersky/OrchardCoreWorkflowsExamples/blob/main/WorkflowWordpressImport.md

**Comment by hyzx86** [comment]
Use Workflow , https://github.com/EasyOC/EasyOC/issues/1
![image](https://user-images.githubusercontent.com/15613121/185099012-17759620-08b3-4568-aed4-cce8f5f822da.png)

**Reply by hyzx86** [reply]
I use the form submission of the workflow to accept an uploaded excel

In the filter area, it will be resolved to datatable

In the script area, the data can be processed directly and finally saved in the `excelData` property

Since the relevant nuget package has not been released, you can refer to my source code to build your business

**Reply by hyzx86** [reply]
In addition, if you do not intend to upload from the page, you will be asked to enter a physical path or a shared directory after deselecting

**Reply by hyzx86** [reply]
this is my upload page

 
![image](https://user-images.githubusercontent.com/15613121/185126386-f7797610-923e-4415-bdbd-b8f4658b0159.png)

**Comment by sbiaudet** [comment]
@Skrypt 

I've recently seen this video https://orcharddojo.net/blog/content-items-excel-file-import-export-add-admin-relative-url-for-admin-menu-nodes-this-week-in-orchard-16-12-2022.

Do you known when it will be available ?

**Reply by Skrypt** [reply]
@MikeAlhayek I think you will need to ask Mike Alhayek for this one. I'm not even sure it is part of Lombiq contrib modules.

**Reply by MikeAlhayek** [reply]
@sbiaudet there is no current plans to bring that code to OC. This code is part of a large project I am working on. Although the code import perfectly today, there are few things that I need to do to improve the functionally for importing larger files.

**Reply by Skrypt** [reply]
You can use Import/Export feature to do that right now and it supports larger files. Only thing you need to do is convert your data to proper JSON.

**Reply by MikeAlhayek** [reply]
the project currently, allows to import large file, but instead of processing the data using a background worker it display them on the UI for you to fix any issues with the data. Like if you are tying to create content with missing data. Eventually, this will be designed to process the data in using a background task.  But yes, if you are able to create a valid json from your file somehow, you can use the current import feature.

**Reply by sbiaudet** [reply]
@MikeAlhayek if you want we can collaborate with you on this feature ?

**Reply by MikeAlhayek** [reply]
@sbiaudet I should be able to get back to improve this in a month or so. I just release a `Queues` module (https://www.nuget.org/packages/CrestApps.Queues) that allows us to queue anything like sending emails or notifications in the background. I think this module will help to import large files by queuing the records into the queue. I'll update you when I work on the import from excel.

**Reply by sbiaudet** [reply]
Hi @MikeAlhayek, do you have any news about the import from excel ? My proposition to work with you on it still’s available.

**Reply by MikeAlhayek** [reply]
@sbiaudet unfortunately I have not had much time lately to work on this. Some paying jobs came along and took higher presence. It's still on my to do list.

**Reply by sbiaudet** [reply]
Hi @MikeAlhayek, thanks for your feedback. For a paying project, I should import 18 000 Content Items from a CSV. So I'm very interested to this feature. 

Can we help you ?? I can free a resource on my team to work on it under your validation. I contact you in private message.

**Reply by nickaynes** [reply]
@sbiaudet I can help with this.Please dm me if interested

**Reply by MikeAlhayek** [reply]
@sbiaudet new PR for this https://github.com/OrchardCMS/OrchardCore/pull/14630 still work in progress with no due date to completion.

**Reply by cocoke** [reply]
Hello @MikeAlhayek, I'm currently working with @sbiaudet and I'm taking over the integration of this functionality on our platform. We're very interested in this module and we can take part in its development if you wish.

I've already started testing a few features on my own by cloning your repo. Is it or will it be possible to manually select the column mapping from a csv/xls to a content item?

**Reply by MikeAlhayek** [reply]
@cocoke let's shift communication into #14630 PR. You can see my response https://github.com/OrchardCMS/OrchardCore/pull/14630#issuecomment-1836427644

**Comment by hyzx86** [comment]
@sbiaudet 

Batch import function ScriptQuery I use here,
https://github.com/EasyOC/EasyOC/blob/main/src/Modules/EasyOC.Excel/Scripting/ExcelMethodsProvider.cs

Using Javascript to organize the matching between imported data and content items, you can read data from many places whether it's Excel or performing external database queries

The problems encountered during the development process are as follows:
1. When a large amount of data is written, the Document table will be occupied for a long time, so page import is needed, of course, this may not be able to use database transactions

2. Some unnecessary queries need to be masked. Many Contenthandlers in OC will be triggered when creating content items.
The problem is that some handlers need to query a single content item from the database. So I filtered out some events in the custom import function


relate: https://github.com/OrchardCMS/OrchardCore/issues/13071
and : https://github.com/OrchardCMS/OrchardCore/issues/10858
